File: parse.py
#!/bin/python

JSON_WHITESPACE = [ ' ', '\n', '\t' ]
JSON_SYNTAX = [ '[', ']', '{', '}', ':', ',' ]
JSON_QUOTE = '"'
JSON_NUMBER_CHARS = [ str(d) for d in range(0, 10) ]

def lex_null(inputstr):
    json_null = 'null'
    json_null_len = len(json_null)

    if len(inputstr) >= json_null_len and inputstr[:json_null_len] == json_null:
        rest = inputstr[json_null_len:]
        return None, rest

def lex_bool(inputstr):
    json_true = 'true'
    json_false = 'false'

    json_true_len = len(json_true)
    json_false_len = len(json_false)

    # Check if string is 'true'
    if len(inputstr) >= json_true_len and inputstr[:json_true_len] == json_true:
        rest = inputstr[json_true_len:]
        return True, rest

    # Check if string is 'false'
    if len(inputstr) >= json_false_len and inputstr[:json_false_len] == json_false:
        rest = inputstr[json_false_len:]
        return False, rest

    # string is neither 'true' nor 'false'
    return None, inputstr

def lex_number(inputstr):
    json_number = ''

    # If first c is a number keep adding all following numbers
    for c in inputstr:
        if c in JSON_NUMBER_CHARS:
            json_number += c
        else:
            break

    # If we did not find any number characters
    if not len(json_number):
        return None, inputstr

    # Remove number characters from input
    rest = inputstr[len(json_number):]

    # Return number either as float or as int
    if '.' in json_number:
        return float(json_number), rest
    else:
        return int(json_number), rest

    return None, inputstr

def lex_string(inputstr):
    json_str = ''

    # If input does not start with a quote it is not a string
    if inputstr[0] != JSON_QUOTE:
        return None, inputstr

    # Remove starting quote from input
    inputstr = inputstr[1:]

    # Everything until the closing quote is the json string
    for c in inputstr:
        if c == JSON_QUOTE:
            # Remove string characters and end quote from input
            rest = inputstr[len(json_str) + 1:]
            return json_str, rest
        else:
            json_str += c

    raise Exception('Unexpected end of input wile lexing string')
    #return None, string

def lex(string):
    print("lex()")

    tokens = []

    while len(string):
        # Check if there is a string
        json_string, string = lex_string(string)
        if json_string is not None:
            print('Found string: ' +json_string)
            tokens.append(json_string)
            continue

        # Check if there is a number
        json_number, string = lex_number(string)
        if json_number is not None:
            print('Found number: ' +str(json_number))
            tokens.append(json_number)
            continue

        # Check if there is a bool
        json_bool, string = lex_bool(string)
        if json_bool is not None:
            print('Found bool: ' +str(json_bool))
            tokens.append(json_bool)
            continue

        # Check if there is a NULL value
        json_null, string = lex_bool(string)
        if json_null is not None:
            tokens.append(json_null)
            print('Found NULL: ' +str(json_null))
            continue

        # Skip whitespaces
        if string[0] in JSON_WHITESPACE:
            string = string[1:]
        elif string[0] in JSON_SYNTAX:
            print('Found syntax char: ' +str(string[0]))
            tokens.append(string[0])
            string = string[1:]
        else:
            raise Exception('Unexpected character ' +str(string[0]))


    if tokens is not None:
        print(tokens)


def main():
    print("main()")

    with open("sprites/box.json", "r") as f:
        #print("opened file '" + str(f.name) +"'")
        content = f.read()
        #print("read file '" + str(f.name) +"'")

    #content = '{"foo": [1, 2, {"bar": 2}]}'
    content = """
        {
            "foo":
                [ 1,
                  2,
                  {
                      "bar": 2
                  }
                ]
        },
        {
            "foo": "bazboo"
        },
        ?
    """
    print("content:\n" + content)
    print("\n")
    lex(content)



if __name__ == '__main__':
    main()

File: parse.cpp

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#include <string.h>

#include <vector>
#include <string>



class JsonItem {
    private:
        JsonItem();
        ~JsonItem();
};

class JsonObject : public JsonItem {
    private:
        std::vector<JsonItem> items;

};



enum JsonType {
    JSON_STRING,
    JSON_INTEGER,
    JSON_BOOLEAN,
    JSON_NULL,
    JSON_SPECIAL
};

struct JsonToken {
    void *data;
    enum JsonType type;
};

static inline bool parse_array(std::vector<JsonToken> *tokens) {
    printf("parse_array()\n");
}

static inline bool parse_object(std::vector<JsonToken> *tokens) {
    printf("parse_object()\n");
}

bool parse(std::vector<JsonToken> *tokens) {
    JsonToken start_t = (*tokens)[0];
    if (start_t.type != JSON_SPECIAL) {
        printf("error - first char is not valid\n");
        return false;
    }

    char start_c = *(char *)start_t.data;
    if (start_c == '{') {
        parse_object(tokens);
    } else if (start_c == '[') {
        parse_array(tokens);
    }

    return false;
}

static inline bool lex_special(char **input, std::vector<JsonToken> *tokens) {
    bool found = false;

    if (*input[0] == '[' || *input[0] == ']') {
        found = true;
    } else if (*input[0] == '{' || *input[0] == '}') {
        found = true;
    } else if (*input[0] == ':' || *input[0] == ',') {
        found = true;
    }

    if (!found) {
        return false;
    }

    char *special_pointer = (char *)malloc(sizeof(char));
    *special_pointer = *input[0];
    JsonToken token = { special_pointer, JSON_SPECIAL };
    tokens->push_back(token);

    (*input)++;
    return true;
}

static inline bool lex_null(char **input, std::vector<JsonToken> *tokens) {
    const char* json_null = "null";
    const size_t json_null_len = strlen(json_null);

    if (strlen(*input) > json_null_len) {
        if (strncmp(*input, json_null, json_null_len) == 0) {
            *input += json_null_len;

            JsonToken token = { NULL, JSON_NULL };
            tokens->push_back(token);
            return true;
        }
    }

    // Did not find null
    return false;
}

static inline bool lex_boolean(char **input, std::vector<JsonToken> *tokens) {
    // Check if there is true boolean
    const char* json_true = "true";
    const size_t json_true_len = strlen(json_true);
    if (strlen(*input) > json_true_len) {
        if (strncmp(*input, json_true, json_true_len) == 0) {
            // Found boolean
            (*input) += json_true_len;

            // Add token
            bool *bool_pointer = (bool *)malloc(sizeof(bool));
            *bool_pointer = true;
            JsonToken token = { bool_pointer, JSON_BOOLEAN };
            tokens->push_back(token);
            return true;
        }
    }

    // Check if there is false boolean
    const char* json_false = "false";
    const size_t json_false_len = strlen(json_false);
    if (strlen(*input) > json_false_len) {
        if (strncmp(*input, json_false, json_false_len) == 0) {
            // Found boolean
            (*input) += json_false_len;

            // Add token
            bool *bool_pointer = (bool *)malloc(sizeof(bool));
            *bool_pointer = false;
            JsonToken token = { bool_pointer, JSON_BOOLEAN };
            tokens->push_back(token);
            return true;
        }
    }

    // Did not find boolean value
    return false;
}

static inline bool lex_number(char **input, bool* error, std::vector<JsonToken> *tokens) {
    // Keep track if number is a float or an int
    bool is_float = false;
    bool is_signed = false;

    // If first char is a digit keep adding digits to number
    std::string json_number = "";
    while (strlen(*input) != 0) {
        char c = (*input)[0];
        // Legal chars for numbers are digits 0-9 minus sign and period
        // TODO: Add support for e
        if ((c > 0x2F && c < 0x3A) || c == '.' || c == '-') {
            // Add
            json_number += c;
            (*input)++;

            // Check if float
            if (c == '.') {
                if (is_float) {
                    // Error: Number contains two periods
                    fprintf(stderr,
                            "Unexpected char %c when parsing number %s\n",
                            c, json_number.c_str());
                    *error = true;
                    return false;
                } else {
                    is_float = true;
                }
            }

            // Check if signed
            if (c == '-') {
                if (is_signed) {
                    // Error: Number contains two minus signs
                    fprintf(stderr,
                            "Unexpected char %c when parsing number %s\n",
                            c, json_number.c_str());
                    *error = true;
                    return false;
                } else {
                    is_signed = true;
                }
            }
        } else {
            // Char is not legal number char -> indicates end of number
            break;
        }
    }

    if (strlen(json_number.c_str()) == 0) {
        // Did not find a number
        return false;
    }

    int inum;
    if (is_float) {
        double dnum = atof(json_number.c_str());

        // Try to convert to int
        inum = (int)dnum;
        if (dnum - inum != 0) {
            fprintf(stderr, "No support for floating point numbers like %f!\n", dnum);
            *error = true;
            return false;
        }
    } else {
        long lnum = atol(json_number.c_str());
        if (lnum > INT_MAX) {
            fprintf(stderr, "Number %ld is too big for int value!\n", lnum);
            *error = true;
            return false;
        }
        inum = (int)lnum;
    }

    int *num_pointer = (int *)malloc(sizeof(int));
    *num_pointer = inum;
    JsonToken token = { num_pointer, JSON_INTEGER };
    tokens->push_back(token);
    return true;
}

static inline bool lex_string(char **input, bool *error, std::vector<JsonToken> *tokens) {
    const char JSON_QUOTE = '"';

    // If first character is not a quote input is not a string
    if ((*input)[0] != JSON_QUOTE) {
        return false;
    } else {
        // Remove starting quote from input
        (*input)++;
    }

    // Everything until the closing quote is part of json string
    std::string json_str = "";
    while (strlen(*input) != 0) {
        char c = (*input)[0];

        if (c == JSON_QUOTE) {
            // Found closing quote -> string is complete
            (*input)++;

            // Add token
            char *out = (char *)malloc(strlen(json_str.c_str()) + 1);
            out = strcpy(out, json_str.c_str());
            JsonToken token = { out, JSON_STRING };
            tokens->push_back(token);
            return true;
        } else {
            // Add character to string
            json_str += c;
            (*input)++;
        }
    }

    // Error: Did not find a closing quote
    *error = true;
    fprintf(stderr, "Unexpected end of input while lexing string\n");
    return false;
}

static inline bool lex_all(char *input, std::vector<JsonToken> *tokens) {
    char *data = input;

    while (strlen(data) != 0) {
        // Check if there is a string
        {
            bool error = false;
            bool added = lex_string(&data, &error, tokens);

            if (error) {
                fprintf(stderr, "Parsing failed\n");
                return false;
            }

            if (added) {
                continue;
            }
        }

        // Check if there is a number
        {
            bool error = false;
            bool added = lex_number(&data, &error, tokens);

            if (error) {
                fprintf(stderr, "Parsing failed\n");
                return false;
            }

            if (added) {
                continue;
            }
        }

        // Check if there is a boolean
        if (lex_boolean(&data, tokens)) {
            continue;
        }

        // Check if there is a NULL value
        if (lex_null(&data, tokens)) {
            continue;
        }

        // Check if there is a syntax char
        if (lex_special(&data, tokens)) {
            continue;
        }

        // Skip whitespaces
        if (data[0] == ' ' || data[0] == '\n' /** TODO: ADD **/ ) {
            data++;
            continue;
        }

        fprintf(stderr, "  Found unexpected character %c\n", data[0]);
        return false;
    }

    return true;
}

bool lex(char *input, std::vector<JsonToken> *tokens) {
    if (!lex_all(input, tokens)) {
        return false;
    }

    printf("Lexing got %d tokens\n", (int)tokens->size());
    for (size_t i = 0; i < tokens->size(); i++) {
        JsonToken t = (*tokens)[i];
        switch (t.type) {
            case JSON_STRING:
                printf("[%02d]  str: '%s'\n", (int)i, (char *)t.data);
                break;
            case JSON_INTEGER:
                printf("[%02d]  int: '%d'\n", (int)i, *(int *)t.data);
                break;
            case JSON_BOOLEAN:
                printf("[%02d]  int: '%s'\n", (int)i, *(bool *)t.data == true ? "true" : "false");
                break;
            case JSON_NULL:
                printf("[%02d]  null\n", (int)i);
                break;
            case JSON_SPECIAL:
                printf("[%02d]  str: '%c'\n", (int)i, *(char *)t.data);
                break;
            default:
                printf("[%02d]\n", (int)i);
                break;
        }
    }

    return true;
}

int main() {
    // Input
    char content[] = "{\"foo\": [1, 2, {\"bar\": 2}], \"visible\": false},{ \"others\": null, \"number\": 5.0, \"one\": 100 }";

    // Lex json
    std::vector<JsonToken> tokens;
    lex(content, &tokens);

    printf("Lexing complete\n");
    printf("%s\n", content);

    // Parse tokens
    parse(&tokens);

    // Free tokens
    for (size_t i = 0; i < tokens.size(); i++) {
        free(tokens[i].data);
    }

    return 0;
}

